### ISSUE TITLE
[Proposal] submodes

### ISSUE THREAD

**Issue by mg979** [neovim/neovim#16313]
Ideas for a `submode` implementation, inspired from https://github.com/neovim/neovim/issues/16263#issuecomment-963083127

A submode would be defined similarly to how `augroup` is defined.

    submode[!] <name>       start defining a submode called <name>
    submode END             end submode definition

`!` means that a previously defined submode with the same name is redefined,
otherwise redefining the same submode is an error.

Inside a `submode` block only mappings, abbreviations, commands and
autocommands can be defined (everything else is an error).

-------------------------------------------------------------------------------

For example:

```vim
submode! mysubmode
    nnoremap bla blabla
    inoreabbrev bla blabla
    autocmd BufLeave SubModeEnd
    command SubModeEnd call submodeend('mysubmode', 'buffer')
submode END
```

Note that `autocmd` has no pattern, because it has submode scope.

Autocommands wouldn't need to be put inside a group, because they would have
the submode scope. No need of `au!` either, because a submode could not be
redefined (`submode!` would clear everything inside the previously defined
submode). Also commands wouldn't need `!` for the same reason.

-------------------------------------------------------------------------------

You would enable it with:

    submodestart(<name>[, <scope>])

where `<scope>` would be global by default, or it can be `buffer` or `window`.

You would end it with:

    submodeend(<name>[, <scope>])

A mode that is enabled for a scope, would need to be disabled for that scope:

    call submodestart('mysubmode', 'buffer')
    call submodeend('mysubmode')                " ERROR: no global submode 'mysubmode'
    call submodeend('mysubmode', 'window')      " ERROR: no window submode 'mysubmode'

-------------------------------------------------------------------------------

Different submodes could be enabled at the same time in the same context, then
if two submodes define conflicting commands/mappings, the last one would win.

To make them stackable, each window/buffer would have a list of enabled
submodes, plus there would be a global list for the globally enabled submodes.
Each submode that is enabled in the current context is checked for a matching
command/mapping/autocommand. Submodes would be checked in reverse order, so the
last enabled submode would be checked first.

**Comment by zeertzjq** [comment]
"submode" probably already has a meaning?

**Comment by mg979** [comment]
None that I know of.

**Comment by clason** [comment]
I think the original proposal by @bfredl was to introduce "namespaces" for mappings and events, which aligns more closely to **Neo**vim's API. ("Submodes" come from https://github.com/kana/vim-submode and are inspired by Emacs, but we should strive for internal consistency.)

Similarly, the API design should be done in this order:
1. internal `nvim_XXX` API
2. Lua API
3. Vim script

**Comment by mg979** [comment]
A `submode` is just another name for this kind of namespaces, as I see it. Once you do it for mappings, I think it makes sense to allow also autocommands and commands inside of them, since they can also be buffer-local and can benefit from having a submode scope. @bfredl was referring to emacs minor modes, judging from the [emacs documentation](https://www.gnu.org/software/emacs/manual/html_node/emacs/Minor-Modes.html) they allow more than just mappings (I never used emacs though, so I'm not sure).

The API I described is just a simplification to illustrate the concept.

**Comment by bfredl** [comment]
Well that's the thing with namespaces: they could be used for anything, though we incrementally implement one feature to them at a time. Though autocmds already have autocmd groups (and filters), so namespaces for mappings is the lowest hanging fruit here. for user ex commands plugins should just use properly disambiguated names anyway.

**Comment by mg979** [comment]
Has some work been done to implement namespaces already?

**Comment by bfredl** [comment]
No.

**Comment by mg979** [comment]
Well I only proposed ideas for a possible implementation, then I don't know if you have already something else in mind about them.

**Comment by neovim-discourse** [comment]
This issue has been mentioned on **Neovim Discourse**. There might be relevant details there:

https://neovim.discourse.group/t/disabling-and-re-enabling-groups-of-mappings/1721/2

**Comment by JamesWidman** [comment]
I like how nvim_namespace_start('foo', 'buffer') would allow you to enable mappings for a buffer without tying the <buffer> modifier to many individual map commands.  A plugin author could put a bunch of mappings into one or more namespaces that their plugin defines, and then let the user decide how they want those namespaces to be scoped.

**Comment by JamesWidman** [comment]
If no one else has started work on this, I would like to take a crack at the implementation!

I have experience with C, but this would be my first time working with the nvim src.  Can anyone recommend regions of the code that I should become familiar with (in general and/or regions that are relevant to this feature)?

**Comment by zeertzjq** [comment]
A lot of code related to mappings, abbreviations and input are in `src/nvim/getchar.c`.

**Comment by JamesWidman** [comment]
in case anyone is interested: i got familiar with the relevant code in getchar.c and related regions (thanks zeertzjq!) but ultimately timed out, partly because other issues took priority.

that said, i did put some work into refactoring, which may be helpful in case anyone else needs to work in this area; e.g. buf_do_map() performs 3 very different tasks with maphashes, and when those tasks are split into 3 different functions (which in my patch i've named 'activate_keymap()', 'deactivate_keymap()', and 'display_active_keymaps()'), those new functions are each very short (about a page long on average) and straightforward. So if anyone is interested in that, let me know!

**Comment by bfredl** [comment]
@JamesWidman sure. you can always publish the code as a draft PR, then we can discuss how to utilize it best moving forward.

**Comment by JamesWidman** [comment]
[done](https://github.com/neovim/neovim/pull/17894)! (well, to the extent that a draft is "done")

**Comment by justinmk** [comment]
> A submode would be defined similarly to how `augroup` is defined.

Please, god, no. More vimscript syntax is the last thing we need. 

This can be achieved with APIs and functions, without new syntax.

**Comment by mg979** [comment]
@justinmk ok it was just an initial idea to make it understandable

**Comment by JamesWidman** [comment]
@justinmk agreed! (fwiw the draft pr doesn't even add a new feature, but if/when someone does that, then for sure we want to expose it only through an api)

**Comment by max397574** [comment]
couldn't we get something coming really close to this if we implement the idea of using the return value of `vim.on_key()`?

**Comment by justinmk** [comment]
Tracked in https://github.com/neovim/neovim/issues/992#issuecomment-2577779050
